### ISSUE TITLE
Unable to serialize Gamepad related struct & enums

### ISSUE THREAD

**Issue by PradeepKumarRajamanickam** [bevyengine/bevy#523]
Was trying out bevy 0.2.0 crate. I am having issue with serde and Gamepad.

Log
```
error[E0277]: the trait bound `bevy::prelude::GamepadButton: axis::_::_serde::Serialize` is not satisfied
  --> src/bindings.rs:10:5
   |
10 |     #[serde(default, rename = "GamepadButtons")]
   |     ^ the trait `axis::_::_serde::Serialize` is not implemented for `bevy::prelude::GamepadButton`
   |
   = note: required because of the requirements on the impl of `axis::_::_serde::Serialize` for `std::collections::HashMap<bevy::prelude::GamepadButton, std::string::String>`
   = note: required by `axis::_::_serde::ser::SerializeStruct::serialize_field`

error[E0277]: the trait bound `bevy::prelude::GamepadButton: axis::_::_serde::Deserialize<'_>` is not satisfied
  --> src/bindings.rs:10:5
   |
10 |     #[serde(default, rename = "GamepadButtons")]
   |     ^ the trait `axis::_::_serde::Deserialize<'_>` is not implemented for `bevy::prelude::GamepadButton`
   |
   = note: required because of the requirements on the impl of `axis::_::_serde::Deserialize<'_>` for `std::collections::HashMap<bevy::prelude::GamepadButton, std::string::String>`
   = note: required by `axis::_::_serde::de::SeqAccess::next_element`

error[E0277]: the trait bound `bevy::prelude::GamepadButton: axis::_::_serde::Deserialize<'_>` is not satisfied
  --> src/bindings.rs:10:5
   |
10 |     #[serde(default, rename = "GamepadButtons")]
   |     ^ the trait `axis::_::_serde::Deserialize<'_>` is not implemented for `bevy::prelude::GamepadButton`
   |
   = note: required because of the requirements on the impl of `axis::_::_serde::Deserialize<'_>` for `std::collections::HashMap<bevy::prelude::GamepadButton, std::string::String>`
```

**Comment by Moxinilian** [comment]
It seems like the Serialize and Deserialize trait are not implemented for those stricts as they should be. Fixing this is just a matter of adding the derive macros like it is done for keyboard values in `bevy_input`.

**Comment by PradeepKumarRajamanickam** [comment]
Opened a PR. Hope thats the right fix.

**Comment by PradeepKumarRajamanickam** [comment]
Hmm... even though the compiler errors are there is a runtime panic.

`    Finished dev [unoptimized + debuginfo] target(s) in 11.82s
     Running `target/debug/examples/with_code RUST_BACKTRACE=1`
Input Map Binding In Code Example
thread 'Compute Task Pool (6)' panicked at 'called `Result::unwrap()` on an `Err` value: "Failed to generate json"', example/binding_in_code.rs:56:53
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'Compute Task Pool (7)' panicked at 'task has failed', /Users/pradeepkumar/.cargo/registry/src/github.com-1ecc6299db9ec823/async-task-4.0.0/src/task.rs:368:45
thread 'main' panicked at 'task has failed', /Users/pradeepkumar/.cargo/registry/src/github.com-1ecc6299db9ec823/async-task-4.0.0/src/task.rs:368:45
thread 'main' panicked at 'task thread panicked while executing: Any', /Users/pradeepkumar/works/rust/bevy_fork/crates/bevy_tasks/src/task_pool.rs:72:18
stack backtrace:
   0:        0x10711677e - backtrace::backtrace::libunwind::trace::he0a1acffd944ce3f
                               at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1:        0x10711677e - backtrace::backtrace::trace_unsynchronized::ha35c8b21934b5ff7
                               at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2:        0x10711677e - std::sys_common::backtrace::_print_fmt::h5e52d5d250a5076b
                               at src/libstd/sys_common/backtrace.rs:78
   3:        0x10711677e - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h24bb64d98a7e25d6
                               at src/libstd/sys_common/backtrace.rs:59
   4:        0x10713e5ac - core::fmt::write::h8fdc9cddb01cd8b2
                               at src/libcore/fmt/mod.rs:1076
   5:        0x107112469 - std::io::Write::write_fmt::hcc3030013983bab6
                               at src/libstd/io/mod.rs:1537
   6:        0x107118c45 - std::sys_common::backtrace::_print::hb08b3b51b5188d8a
                               at src/libstd/sys_common/backtrace.rs:62
   7:        0x107118c45 - std::sys_common::backtrace::print::h97998656489491bc
                               at src/libstd/sys_common/backtrace.rs:49
   8:        0x107118c45 - std::panicking::default_hook::{{closure}}::h95817712c5ff0736
                               at src/libstd/panicking.rs:198
   9:        0x107118982 - std::panicking::default_hook::h34e085f4e0b1062d
                               at src/libstd/panicking.rs:217
  10:        0x1071191a5 - std::panicking::rust_panic_with_hook::haf571858f996ac45
                               at src/libstd/panicking.rs:526
  11:        0x107118d72 - rust_begin_unwind
                               at src/libstd/panicking.rs:437
  12:        0x10716534f - core::panicking::panic_fmt::h11676ba6a846d9f4
  13:        0x107165255 - core::option::expect_none_failed::h312e48bacb63d8ec
  14:        0x1070a53b6 - core::result::Result<T,E>::expect::hc5182e8e4dcf8680
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/result.rs:963
  15:        0x10709a335 - <bevy_tasks::task_pool::TaskPoolInner as core::ops::drop::Drop>::drop::h06ed3b111bc19c0f
                               at /Users/pradeepkumar/works/rust/bevy_fork/crates/bevy_tasks/src/task_pool.rs:70
  16:        0x1070986e5 - core::ptr::drop_in_place::h5cad8f68798a91c3
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/ptr/mod.rs:184
  17:        0x107003a32 - alloc::sync::Arc<T>::drop_slow::h3ee5bd66b0ccd2a7
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/liballoc/sync.rs:864
  18:        0x1070170fa - <alloc::sync::Arc<T> as core::ops::drop::Drop>::drop::h30ac03b2bc8c850c
                               at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.6.2/src/raw/mod.rs:1122
  19:        0x107015891 - core::ptr::drop_in_place::h3c9f8880b8bf8771
                               at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.6.2/src/raw/mod.rs:1122
  20:        0x10701638e - core::ptr::drop_in_place::hb180a36976313219
                               at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.6.2/src/raw/mod.rs:1122
  21:        0x107015951 - core::ptr::drop_in_place::h43717040449ffc1e
                               at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.6.2/src/raw/mod.rs:1122
  22:        0x106fc67e1 - core::ptr::mut_ptr::<impl *mut T>::drop_in_place::h2a0e62436655527b
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/ptr/mut_ptr.rs:846
  23:        0x106fdd65e - bevy_hecs::archetype::TypeInfo::of::drop_ptr::ha7f46a6ce43b6eea
                               at /Users/pradeepkumar/works/rust/bevy_fork/crates/bevy_ecs/hecs/src/archetype.rs:487
  24:        0x1070da306 - bevy_hecs::archetype::Archetype::clear::h3584f639d12ef707
                               at /Users/pradeepkumar/works/rust/bevy_fork/crates/bevy_ecs/hecs/src/archetype.rs:85
  25:        0x107101a55 - <bevy_hecs::archetype::Archetype as core::ops::drop::Drop>::drop::h0e8838f27404c524
                               at /Users/pradeepkumar/works/rust/bevy_fork/crates/bevy_ecs/hecs/src/archetype.rs:419
  26:        0x107100815 - core::ptr::drop_in_place::h47567f6a06dde15e
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/ptr/mod.rs:184
  27:        0x106fd3205 - core::ptr::drop_in_place::he83e66f6ed1e8ef9
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/ptr/mod.rs:184
  28:        0x106fd2978 - core::ptr::drop_in_place::h88c77559582f980d
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/ptr/mod.rs:184
  29:        0x106fe7eb1 - core::ptr::mut_ptr::<impl *mut T>::drop_in_place::ha9ef6cbda7367b68
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/ptr/mut_ptr.rs:846
  30:        0x106fc734e - hashbrown::raw::Bucket<T>::drop::hcde31f9f14a7d36a
                               at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.6.2/src/raw/mod.rs:307
  31:        0x106fd3ab1 - <hashbrown::raw::RawTable<T> as core::ops::drop::Drop>::drop::hd7dd216a4cdd8492
                               at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.6.2/src/raw/mod.rs:1041
  32:        0x106fd3261 - core::ptr::drop_in_place::hed506ba7d4dd4e71
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/ptr/mod.rs:184
  33:        0x106fd20d8 - core::ptr::drop_in_place::h26434631be61c9d5
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/ptr/mod.rs:184
  34:        0x106fd2c81 - core::ptr::drop_in_place::ha79ceda7b1fb857d
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/ptr/mod.rs:184
  35:        0x106fd3151 - core::ptr::drop_in_place::hd8018e28522518b0
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/ptr/mod.rs:184
  36:        0x106fd226c - core::ptr::drop_in_place::h2dff2fe8a6115509
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/ptr/mod.rs:184
  37:        0x106fd9d37 - bevy_app::app::App::run::hb5e9dd9912f5e638
                               at /Users/pradeepkumar/works/rust/bevy_fork/crates/bevy_app/src/app.rs:83
  38:        0x106fe6b80 - bevy_app::app_builder::AppBuilder::run::hc1706a98eea241a5
                               at /Users/pradeepkumar/works/rust/bevy_fork/crates/bevy_app/src/app_builder.rs:45
  39:        0x105c690e1 - with_code::main::h3f492b3fc439b815
                               at example/binding_in_code.rs:9
  40:        0x105c69c4e - std::rt::lang_start::{{closure}}::he05f1891c8136c17
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libstd/rt.rs:67
  41:        0x10711946f - std::rt::lang_start_internal::{{closure}}::h61757db93e0a7614
                               at src/libstd/rt.rs:52
  42:        0x10711946f - std::panicking::try::do_call::h908b994d27b7ba54
                               at src/libstd/panicking.rs:348
  43:        0x10711946f - std::panicking::try::heb108336153d9e3c
                               at src/libstd/panicking.rs:325
  44:        0x10711946f - std::panic::catch_unwind::h5937a18f6925fee9
                               at src/libstd/panic.rs:394
  45:        0x10711946f - std::rt::lang_start_internal::hcd84a36052901671
                               at src/libstd/rt.rs:51
  46:        0x105c69c31 - std::rt::lang_start::h595cb26745464f81
                               at /Users/pradeepkumar/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libstd/rt.rs:67
  47:        0x105c69be2 - main
thread panicked while panicking. aborting.
Illegal instruction: 4`

@Moxinilian Any suggestions for this?

**Comment by Moxinilian** [comment]
Could you send the error message in a code block instead (using triple backquotes instead of single backquotes), I'm having a really hard time deciphering the message without line breaks. But the fact that it panics while panicking is really not good.

**Comment by PradeepKumarRajamanickam** [comment]
Done I have update the message with triple backquotes.

**Comment by cart** [comment]
closed by #583
